File: src/reporter.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Verb {
    Translating,
    Revisiting,
    Validating,
    Checking,
    Ok,
    Stale,
    Missing,
    Removed,
    Skipped,
    Translated,
    Revisited,
    Cleaned,
    Created,
    Updated,
    Summary,
    Info,
    DryRun,
}

impl Verb {
    pub fn as_str(&self) -> &'static str {
        match self {
            Verb::Translating => "Translating",
            Verb::Revisiting => "Revisiting",
            Verb::Validating => "Validating",
            Verb::Checking => "Checking",
            Verb::Ok => "Ok",
            Verb::Stale => "Stale",
            Verb::Missing => "Missing",
            Verb::Removed => "Removed",
            Verb::Skipped => "Skipped",
            Verb::Translated => "Translated",
            Verb::Revisited => "Revisited",
            Verb::Cleaned => "Cleaned",
            Verb::Created => "Created",
            Verb::Updated => "Updated",
            Verb::Summary => "Summary",
            Verb::Info => "Info",
            Verb::DryRun => "Dry run",
        }
    }
}

pub trait ProgressReporter {
    fn increment(&mut self, label: &str);
    fn done(&mut self);
}

pub trait Reporter: Send + Sync {
    fn log(&self, verb: Verb, message: &str);
    fn step(&self, verb: Verb, current: usize, total: usize, message: &str);
    fn blank(&self);
    fn progress(&self, verb: Verb, total: usize) -> Box<dyn ProgressReporter>;
}

struct NoopProgress;

impl ProgressReporter for NoopProgress {
    fn increment(&mut self, _label: &str) {}
    fn done(&mut self) {}
}

pub struct NoopReporter;

impl Reporter for NoopReporter {
    fn log(&self, _verb: Verb, _message: &str) {}
    fn step(&self, _verb: Verb, _current: usize, _total: usize, _message: &str) {}
    fn blank(&self) {}
    fn progress(&self, _verb: Verb, _total: usize) -> Box<dyn ProgressReporter> {
        Box::new(NoopProgress)
    }
}

pub fn ensure_reporter(reporter: Option<&dyn Reporter>) -> &dyn Reporter {
    static NOOP: NoopReporter = NoopReporter;
    reporter.unwrap_or(&NOOP)
}

File: cli/internal/glossia/reporter.go
package glossia

import (
	"fmt"
	"os"
	"strings"

	"golang.org/x/term"
)

type verb string

const (
	verbTranslating verb = "Translating"
	verbRevisiting  verb = "Revisiting"
	verbValidating  verb = "Validating"
	verbChecking    verb = "Checking"
	verbOk          verb = "Ok"
	verbStale       verb = "Stale"
	verbMissing     verb = "Missing"
	verbRemoved     verb = "Removed"
	verbSkipped     verb = "Skipped"
	verbCleaned     verb = "Cleaned"
	verbCreated     verb = "Created"
	verbUpdated     verb = "Updated"
	verbSummary     verb = "Summary"
	verbInfo        verb = "Info"
	verbDryRun      verb = "Dry run"
)

type Reporter interface {
	Log(v verb, message string)
	Step(v verb, current int, total int, message string)
	Blank()
}

const (
	colorReset  = "\u001b[0m"
	colorBold   = "\u001b[1m"
	colorDim    = "\u001b[2m"
	colorGreen  = "\u001b[32m"
	colorCyan   = "\u001b[36m"
	colorYellow = "\u001b[33m"
	colorRed    = "\u001b[31m"
	colorWhite  = "\u001b[37m"
)

func colorForVerb(v verb) string {
	switch v {
	case verbOk, verbRemoved, verbCleaned, verbCreated, verbUpdated:
		return colorGreen
	case verbTranslating, verbRevisiting, verbValidating, verbChecking:
		return colorCyan
	case verbStale, verbSkipped, verbDryRun:
		return colorYellow
	case verbMissing:
		return colorRed
	default:
		return colorWhite
	}
}

func formatVerb(v verb, useColor bool) string {
	padded := fmt.Sprintf("%12s", v)
	if !useColor {
		return padded
	}

	return strings.Join([]string{colorBold, colorForVerb(v), padded, colorReset}, "")
}

type ConsoleReporter struct {
	useColor bool
}

func NewConsoleReporter(noColor bool) *ConsoleReporter {
	return &ConsoleReporter{
		useColor: !noColor && term.IsTerminal(int(os.Stdout.Fd())),
	}
}

func (r *ConsoleReporter) Log(v verb, message string) {
	_, _ = os.Stdout.WriteString(fmt.Sprintf("%s  %s\n", formatVerb(v, r.useColor), message))
}

func (r *ConsoleReporter) Step(v verb, current int, total int, message string) {
	r.Log(v, fmt.Sprintf("[%d/%d] %s", current, total, message))
}

func (r *ConsoleReporter) Blank() {
	_, _ = os.Stdout.WriteString("\n")
}
